# Part 1: Haskell is Just Like Every Other Language

> I have made this longer than usual because I have not had time to make it shorter.
> - Blaise Pascal

## Comments

Before we get started, Haskell has single line comments:

In [1]:
-- This is a single line comment, similar to C's // or python's #

and multi line comments:

In [2]:
{- This is a multi line comment,
   similar to C's  /* multi-line comments */
-}

We'll also disable linting to prevent IHaskell from complaining when I write things in an awkward way for pedagogical reasons.

In [3]:
:opt no-lint

Now we can dive into the good stuff.

# Built in Data Types

Haskell's set of built in data types is comparable to most languages.

## Unit

The unit type is a type that has a single value, pronounced "unit" but spelled `()`.  This type is not very useful in isolation but is commonly used when you need to provide a value or type somewhere but it doesn't matter to you what the value or type actually is.   A common example is that any function that performs side effects but does not need to return a value will end up being of some type like `IO ()` or `SomethingElse ()`.

(Ok, this one actually isn't very much like other languages, but it's pretty simple and it comes up a lot, so I figured I'd start with it).

In [4]:
u = ()

print u

()

## Booleans

In [5]:
yes = True
no  = False

print yes
print no

True

False

Boolean operations use C style binary operators and python style `not`:

In [6]:
print (yes || no)
print (yes && no)
print (not yes)

True

False

False

## Characters

Characters in Haskell are unicode.

In [7]:
a = 'a'
b = 'b'
c = 'c'
lambda = 'λ'

print a
print b
print c
print lambda

'a'

'b'

'c'

'\955'

The '\955' output for the lambda character is a result of the fact that the `print` function uses Haskell's `show` function to convert it's argument to a string representation (suitable for reading back in) of the value before printing it.  You might think that a string would just display as itself but to be read back in properly it needs to be wrapped in quotes and have any non-ascii characters ecaped.  We'll circle back to this in a moment.

## Numbers

Haskell has a fairly rich (though regrettably poorly designed) numeric tower.  We'll revisit some of the complexities later, but for now, here are some examples of basic integer and floating point arithmetic:

In [8]:
i = 2
j = 3

print i
print j

print (i + j)
print (i * j)

2

3

5

6

In [9]:
d = 3.14159
e = 0.2112

print d
print e

print (d + e)
print (d * e)

3.14159

0.2112

3.3527899999999997

0.663503808

## Lists

Lists in Haskell are (possibly infinite) singly linked lists.  They are widely used both as a data structure and as a control structure.  You can build up a list by using the cons operator (`:`) to prepend values to the empty list (`[]') like so:

In [10]:
xs = 1:2:3:[]

print xs

[1,2,3]

But you can also use the nice syntax sugar similar to many languages by "spelling" the list the same way it was printed above:

In [11]:
ys = [4,5,6]

print ys

[4,5,6]

Haskell provides a rich set of functions for operating on lists and similar data structures, but for now here are a few simple examples:

In [12]:
print (head xs)

1

In [13]:
print (tail ys)

[5,6]

In [14]:
print (xs ++ ys)

[1,2,3,4,5,6]

In [15]:
print (length xs)

3

In [16]:
print (take 2 ys)

[4,5]

In [17]:
print (drop 1 xs)

[2,3]

In [18]:
print (reverse (xs ++ ys))

[6,5,4,3,2,1]

## Strings

...in Haskell are just lists of characters:

In [19]:
cs = ['a', 'b', 'c']

print cs

"abc"

which, similarly, we can spell the natural way:

In [20]:
ds = "def"

print ds

"def"

and operate on with all of the same functions/operators:

In [21]:
print (head cs)
print (cs ++ ds)
print (length cs)

'a'

"abcdef"

3

Now that we have strings in our toolbox we can revisit our unicode rendering issue from above.  Instead of using the `print` function we can use `putStrLn` which avoids the escaping/encoding issue by requiring that its argument is already a string and then renders it unaltered.  Since our `lambda` value above is a single character rather than a string we first have to promote it to a string:

In [22]:
lambdas = [lambda, lambda, lambda]

putStrLn lambdas

λλλ

## Ranges

Haskell provides a range syntax (`[..]`) which allows a shorthand for writing ranges of values.  The basic form is pretty self explanatory:

In [23]:
print [1..10]

[1,2,3,4,5,6,7,8,9,10]

Ranges also work with some other types (specifically any type that is an instance of the `Enum` class):

In [24]:
print [1.5..12.5]
print ['a'..'z']
print [False .. True]  -- need spaces around the `..` operator to distinguish from `Module.reference` syntax.

[1.5,2.5,3.5,4.5,5.5,6.5,7.5,8.5,9.5,10.5,11.5,12.5]

"abcdefghijklmnopqrstuvwxyz"

[False,True]

But ranges also have a few more tricks up their sleeves. 

First, we can specify a second index before the operator to control the size of the increment:

In [25]:
evens  = [2,4..10]
tens   = [10,20..100]
floats = [3.14159,4.567..20]
chars  = ['a', 'c'.. 'r']

print evens
print tens
print floats
print chars

[2,4,6,8,10]

[10,20,30,40,50,60,70,80,90,100]

[3.14159,4.567,5.9924100000000005,7.417820000000001,8.843230000000002,10.268640000000001,11.69405,13.119460000000004,14.544870000000003,15.970280000000002,17.395690000000002,18.821100000000005,20.246510000000004]

"acegikmoq"

Secondly, and more interestingly (IMHO), we can omit the last index to create an infinite list:

In [26]:
ns   = [1..]
odds = [1,3..]

-- print ns    -- If we were to print `ns` here the output would never stop
-- print odds  -- same

You might be wondering why the declaration itself didn't cause the interpreter to hang forever, or what good it is to have an infinite list if you can't do basic things like print it out...  Well, the short answer is:

### Laziness

Due to laziness Haskell will only generate as much of the infinite list as we ask for.  So for example we can use the `take` function that we saw above to create a new list containing only the first `n` elements of our infinite list:

In [27]:
ten    = take 10 ns
twenty = take 20 ns

print ten
print twenty

[1,2,3,4,5,6,7,8,9,10]

[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]

Due to laziness nothing (other than a "thunk") was created when we declared the `ten` and `twenty` values but when we executed `print ten` the thunk was "forced" to produce the `1`, then the `2`, and so on through the ten, but then stopped there since nothing had (yet) asked for anything past the `10`.  You can imagine that at this point the list looked something like `1:2:3:4:5:6:7:8:9:10:thunk` in memory.  On the next line when we executed `print twenty` the `[1..10]` range from the previous evaluation was already in memory but now the `take 20` proceeded past the `10` to hit the `thunk` at the end and continued forcing from there, resulting in the list looking something like `1:2:3:4:5:6:7:8:9:10:11:12:13:14:15:16:17:18:19:20:thunk` in memory.

And of course we can use the same trick to use a portion of our `odds` value that was declared above with a different increment:

In [28]:
print (take 10 odds)

[1,3,5,7,9,11,13,15,17,19]

All values in a list must be of the same type (`Int` in the examples above).  Attempting to write something like `[5, 3.14, 'c', "foo"]` which would work just fine in python will result in an error in Haskell.

## Tuples

In addition to lists, Haskell also has tuples which are of a fixed size but allow different element types:

In [29]:
print ('a','b')
print (1,2,3)
print ("foo", "bar", "baz", "bif")
print ('a', 2, "foo", [("bar", 3, 'b'), ("bif", 4, 'c')])

('a','b')

(1,2,3)

("foo","bar","baz","bif")

('a',2,"foo",[("bar",3,'b'),("bif",4,'c')])

As tempting as it might be to think of tuples as a heterogenous list, it's mostly not usable as one.  For example there's no way to concatenate two tuples into a third bigger tuple, ie. you can't do something like `(1, 'a') ++ ("foo", 5)` to get `(1, 'a', "foo", 5)`.  

It's of course possible to deconstruct two tuples and reconstruct them into a single new tuple manually, but there are numerous other reasons to avoid thinking about tuples this way, for example the unexpected behavior of the `length` function on tuples:

In [30]:
length ('a', 'b')

1

Getting into the details of why this is the case requires too much missing background at this point, so just take my word that you shouldn't think of tuples as lists. :)

## Equality

Haskell uses the common double equals (`==`) operator for equality.  Not every type in Haskell implements equality -- most notablity functions do not, but for any type that supports structural equality (that is, if the value looks the same, it is the same) you can derive equality by just adding `deriving Eq` to your data type definition, which is nice compared to say Java where you end up writing a bunch of redundant `.equals` implementations.

In [31]:
print (1 == 1)
print (1 == 2)

print ('a' == 'a')
print ('b' == 'c')

-- print (1 == 'b') -- will fail because the values must be of the same type

True

False

True

False

## Functions

Functions are where Haskell starts to diverge quite a bit from most mainstream languages, and we'll be digging into that a little more in Part 2, but for now, we can see some simple function definitions that show the commonality with other languages rather than the differences:

In [32]:
add a b = a + b

print (add 1 2)

3

As you can see, function definitions work similary to other languages, though the lack of parentheses may seem a little odd a first.  As it turns out there are good reasons (related lambda calculus which Haskell is based on, currying and the syntax (juxtaposition) for function application) for this decision.

You get used to it pretty quickly but one thing to watch out for is that as a beginner you might end up inadvertantly creating a function which accepts a tuple rather than individual arguments, like so:

In [33]:
sub(a, b) = a - b

This is perfectly valid but you must call it with a tuple like so:

In [34]:
print (sub (8, 3))

5

For functions of two arguments we can actually convert to a function on a 2-tuple and back with the `curry` and `uncurry` (higher order) functions:

In [35]:
uadd = uncurry add
csub = curry sub

print (uadd (1,2))
print (csub 6 2)

3

4

## Higher Order Functions

We've just seen our first example of higher order functions in `curry` and `uncurry` -- these functions each take a function and return a new function.  

If you're familiar with functional approaches in python then you've probably already encountered the higher order functions `map`, `filter` and `reduce`.  Haskell as `map` and `filter` that work the same as python and `foldl'` which is equivalent to `reduce`.

In [36]:
inc x = x + 1

print (map inc [1..3])

[2,3,4]

In [37]:
print (filter even [1..10])

[2,4,6,8,10]

Haskell already has a built in sum function:

In [38]:
print (sum [1..10])

55

But if it didn't we could use `foldl'` to implement it similarly to how you might use `reduce` in python.

NB. there is a `foldl` function (without the trailing tic) which is similar to the version with the tic but is lazy in it's accumulator instead of strict and this is almost never what you want, so if you want to do a left fold you should almost always import the `foldl'` version from `Data.List`:

In [39]:
import Data.List (foldl')

sum' xs = foldl' add 0 xs

print (sum' [1..10])

55

There are also right folds which accomplish the same thing but with different space usage characteristics which can be useful in various situations:

In [40]:
print (foldr add 0 [1..10])

55

When starting out you should probably always reach for `foldl'` first until you have a reason to dig deeper.

The big difference between `foldl'` in Haskell and `reduce` in python is that `foldl'` requires us to pass a default argument when the list is empty.  This is to make the function total so that it can return a sensible result rather than an error even if the list is empty.  In python passing an empty list to `reduce` will result in the following error:

```
>>> reduce(operator.add, [])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: reduce() of empty sequence with no initial value
```

Haskell has `foldl1` which essentially provides this behavior:

In [41]:
foldl1 add [1,2,3]

6

In [42]:
foldl1 add [1]

1

In [43]:
foldl1 add []

: 

This is handy if you happen to know your list is not empty but use of `foldl'` with a default value is much more common in the wild.

## A Brief Digression on `inc`

There are actually two other ways we could write our `inc` function from above.  First, we could partially-apply our `add` function from above:

In [44]:
inca = add 1

print (map inca [1..5])

[2,3,4,5,6]

or we can go a step further with a technique called sectioning where you enclose an application of an operator in parentheses and leave off either side to partially apply it:

In [45]:
incp  = (+1)
incp' = (1+)

print (map incp  [1..3])
print (map incp' [1..3])

[2,3,4]

[2,3,4]

You can also use the section syntax directly inline without defining an intermediate helper function:

In [46]:
print (map (+5) [1..3])

[6,7,8]

While we're on the subject we can also use parentheses to turn an operator into a normal function, so we could have defined our `add` function above more simply like so which we could think of as a section with both sides missing:

In [47]:
add' = (+)

print (add' 1 2)

3

As with the example above you can use the `(+)` syntax directly.  For example our `sum'` function above could be written like so:

In [48]:
sum'' xs = foldl' (+) 0 xs

sum'' [5..10]

45

And, last but not least, we can use back-tics to turn use a regular function as an operator:

In [49]:
print (1 `add` 2)

3

These various techniques can be used to make code more (or less) readable according to taste.

Now that we have functions under our belt we can revisit laziness and see a slightly more involved application of the concept via a common Haskell example:

In [50]:
fibs = 1:1:zipWith (+) fibs (tail fibs)

print (take 10 fibs)

[1,1,2,3,5,8,13,21,34,55]

This example defines the fibonacci sequence by first creating a list consisting of two 1's and then appending to that list a list created by zipping with `(+)` the list itself with the tail of the list itself... so to generate the third value (after the first two 1's) it gets the first value of fibs (which is the first 1) and uses `(+)` to add it to the first value of the tail of the list, which is the second 1, arriving at 2 for the third value in the list (`1:1:2...`).  To calculate the fourth value it proceeds taking combining the 2nd value in the list (the second 1) with the third value in the list (the 2) to get 3, and so on.

## Haskell Lacks the Concept of Null Types

Tony Hoare's [Billion Dollar Mistake](https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare) is not present in Haskell.  I can't tell you how great this (lack of a) feature is just on it's own.

So what do you do if you need to represent something like an optional value that might be handled by `null` (`nil`, `None`, etc) in another language?  

## Maybe

to the rescue!

The maybe type has two constructors `Just` for creating a `Maybe` value that is present and `Nothing` for creating a value which represents the case where no value is not present.  `Maybe` is technically a higher kinded type that needs to be applied to a specific concrete type to create a new specific concrete type, so for example you can apply it to `Int` to get `Maybe Int` or apply it to `Char` to get `Maybe Char`. 

Here are some examples of `Maybe` values:

In [51]:
intMay5 = Just 5

intMayNone = Nothing

charMayC = Just 'c'

charMayNone = Nothing

print intMay5
print intMayNone
print charMayC
print charMayNone

Just 5

Nothing

Just 'c'

Nothing

## User Defined Data Types

We've seen some examples of laziness which might seem pretty strange coming from other languages, and partial application and sectioning might be a little bit strange, but for the most part I don't think anything we've covered is drastically different than what you would expect in other languages.

User defined data types are extremely common and important in Haskell but they start to differ more significantly from other languages, so we'll wrap up this section with a simple example of defining a type that is fairly similar to record types from other lanuages and then dive deeper into the differences as we transition to part 2.

In [52]:
data User = User
  { userId   :: Int
  , username :: String
  , isAdmin  :: Bool
  } deriving (Eq, Ord, Read, Show)

Other than the `deriving` part, hopefully this is fairly intuitive and the correspondence to a `struct` in C, a `namedtuple` in python, or various record types from other languages.

With our `User` type in hand, we can create a values of type `User` like so:

In [53]:
adminUser = User 0 "admin" True
normUser  = User 1 "norm" False

print adminUser
print normUser

User {userId = 0, username = "admin", isAdmin = True}

User {userId = 1, username = "norm", isAdmin = False}

In [54]:
print (adminUser == adminUser)
print (normUser == normUser)
print (adminUser == normUser)

True

True

False

Haskell is not an object oriented language so there's no special support from the language define "methods" that behave the way methods in other languages do (eg with reference to an implicit `this` from Java or explicit `self` from python) but there are many ways to emulate OO in various ways if you really want to.

Typically though we tend to just write functions that operate on values of our types, something like:

In [55]:
import Data.Char ( toUpper )

renderUsername user = map toUpper (username user)

renderUsername adminUser
renderUsername normUser

"ADMIN"

"NORM"

There is an `&` operator which is provided by various packages or can by simply defined like so:

In [56]:
(&) = flip ($)

This operator provides reverse function application which might feel a little more familiar for those coming from an OO language:

In [57]:
adminUser & renderUsername

"ADMIN"

Though you won't see it used in this style too often in the wild.

It's also easy to create a record which holds one or more functions like so:

In [58]:
data Animal = Animal
  { animalName :: String
  , speak      :: Animal -> IO ()
  }

-- We aren't able to derive Eq, Ord, Read, or Show since they are not defined for functions.
-- (Though we could implement our own custom versions if we wanted to)

In [59]:
catSpeak animal = putStrLn (animalName animal ++ " says Meow!")
dogSpeak animal = putStrLn (animalName animal ++ " says Woof!")

garfield = Animal "garfield"  catSpeak
odie = Animal "odie" dogSpeak

speak garfield garfield
odie & speak odie

garfield says Meow!

odie says Woof!

You'll notice that we had to pass garfield twice.  This is because the fields of the record are turned into accessor functions which implicitly take the data type they are declared in as a first parameter but the actual function does not have access to this implicit value -- it is only used to select the slot within the type which the function resides.  Since we wanted to use the name of the animal we had to make it an explicit argument so we could call `animalName` on it.  Thus the actual type of the `speak` function is not `Animal -> IO ()` as we declared it, but `Animal -> (what we declared)` which ends up as `Animal -> Animal -> IO ()` in this case.

As if this awkwardness weren't already enough to put you off of Haskell records, there are even worse problems with records which bite you even if you aren't trying to emulate OO... for example two records can't have fields with the same name:

In [60]:
data Type1 = Type1
  { field1 :: String
  }
  
data Type2 = Type2
  { field1 :: String
  }

: 

There are numerous ways to mitigate this proble.  My favorite solution, lenses, will feature prominently in the next notebook but there are also libraries like `SuperRecord`, language extensions like `DuplicateRecordFields` and so on which address these shortcomings in various ways.

We've already introduced a couple of concepts (like the type signatures in records and especially the `IO ()` type in our `Animal` record) without explaining them and we're going to need stray even more from the beaten path to explain much more, so it looks like it might be time to enter...

# Part 2: Haskell is Unlike Any Other Language 

We've tipped our hand a bit because we had to start showing type signatures in order to demonstrate records but at least until the last few panels you might have been wondering to yourself something like "I thought Haskell was all statically typed like Java but all these examples look dynamic like Python... what gives?"

Haskell is in fact statically typed but unlike Java it has (very good) type inference which allows us to omit type signature in most situations.  So far we've omitted type signatures in every situation except the declaration of our user defined data types where they are required.  Typical Haskell code annotates all top level definitions with type signatures (for various reasons, primary among them being that they serve as very precise executable documentation).

Especially when you've written type signatures for all top level definitions it is rare to need to manually write any other type signatures but you can add signatures to any value or expression in any context if you want.  In rare cases the compiler may not be able to infer the types of some expression and in these cases it's necessary to add them.  There are also good reasons to add non-required type signatures in some other situations -- again primarily when it makes it more obvious or clear what the code is doing.

*Side note: Up until now I've been consistently using `print` to show values for various reasons (eg being able to show two values in one block, etc) but I'll drop this practice for the rest of the document where it's not explicitly necessary.*

In addition to writing type signatures we can see what has been inferred for any given type (here in IHaskell or in the usual Haskell REPL which called `ghci`).

Let's recall some of our values from above:

In [61]:
print yes
print a
print cs
print i
print d
print evens
print floats

True

'a'

"abc"

2

3.14159

[2,4,6,8,10]

[3.14159,4.567,5.9924100000000005,7.417820000000001,8.843230000000002,10.268640000000001,11.69405,13.119460000000004,14.544870000000003,15.970280000000002,17.395690000000002,18.821100000000005,20.246510000000004]

We can check the types of any of these values with `:t` --

In [62]:
:t yes

yes :: Bool

In [63]:
:t a

a :: Char

In [64]:
:t cs

cs :: [Char]

In [65]:
:t i

i :: forall p. Num p => p

In [66]:
:t d

d :: forall p. Fractional p => p

In [67]:
:t evens

evens :: forall a. (Num a, Enum a) => [a]

In [68]:
:t floats

floats :: forall a. (Fractional a, Enum a) => [a]

The first few of these are fairly self explanatory.

For `cs`, recall that a `String` is a just list of `Char`s.  While we can see that this is true, there exists a type alias called `String` that is interchangable with `[Char]` so we can in fact use `String` anywhere `[Char]` is expected (and vice versa):

In [69]:
x = "a string" :: String

:t x

x :: String

Starting with `i` things start to get wierd.

What's happened here is that in the absence of an explicit type signature (or other context that specializes the types) Haskell always infers the most general type possible.

For `i` we might've expected a type like `Int` (the machine-sized integer type) or `Integer` (an unbounded integer type) but we actually got `forall t. Num t => t`.  The `forall` is universal quantification which IHaskell makes explicit, but all type variables are universally quantified in Haskell by default, so in `ghci` this would render as simply `i :: Num t => t` which is at least a little more approachable.  What this says is basically "`i` is of some type `t` which is an instance of the `Num` class" or essentially "`i` is some type of number".

The inferred type for `d` is a little more specific because the value contains a decimal place and that excludes all types of whole numbers (like `Int`s and `Integer`s).

The type for `evens` is still pretty general, but since we used the range operator to construct that list in addition to `Num` it says that `t` must be an instance of the `Enum` class as well (as the range operator requires `Enum`).

The type of `floats` is constrained both to instances of `Fractional` because of the decimal places and `Enum` due to the use of the range operator.

If we wanted to be explicit about types we can add type signatures above the definitions of our values, like so:

In [70]:
int :: Int
int = 2

integer :: Integer
integer = 987654321098765432109876543210

And as expected they will have precisely the types we gave them:

In [71]:
:t int

int :: Int

In [72]:
:t integer

integer :: Integer

We don't have a tuple value handy but we can check the type of one we create on the fly easily enough:

In [73]:
:t (1, 'c')

(1, 'c') :: forall a. Num a => (a, Char)

Hopefully this signature makes sense based on what we've seen above.  Speaking of tuples, let's recall our big goofy composite type from above and see what the inferred type is:

In [74]:
:t ('a', 2, "foo", [("bar", 3, 'b'), ("bif", 4, 'c')])

('a', 2, "foo", [("bar", 3, 'b'), ("bif", 4, 'c')]) :: forall b1 b2. (Num b1, Num b2) => (Char, b1, [Char], [([Char], b2, Char)])

It's a bit complicated but it's really just the logical composition of the various types we're already familiar with.

Speaking of types we're familiar with, we can check the type of values of our custom data types as well:

In [75]:
:t adminUser

adminUser :: User

The type signatures for values of these various data types are pretty straight forward, but we've seen hints of several properties of more advanced signatures in the type we declared for the `speak` function in our `Animal` example.  Recall that the type we wrote was `Animal -> IO ()` and we said that due to the record accessor mumbo jumbo the real type was `Animal -> Animal -> IO ()`.  Let's check and see if we were right:

In [76]:
:t speak

speak :: Animal -> Animal -> IO ()

Great.  We were.

The arrow (`->`) in type signatures indicates a function type.  The `IO ()` in our particular `speak` signature involves purity, parametric polymorphism, and higher kinded types, as well as touching on the dreaded "m" word (monad) so let's forget about all of that for the moment and explore the types of some simpler functions.

We can check the type of any of the other functions we've used so far:

In [77]:
:t print

print :: forall a. Show a => a -> IO ()

This function also involves IO so we'll gloss over that for the moment but you can think of `IO ()` as the return type of a function which performs some side effect (in this case printing to the console) returning no value (similar to `void` in other languages).  But the `Show` class constraint should make more sense now -- the whole type signature basically says "Given a value that is showable I'll print it out on the console".

In [78]:
:t add

add :: forall a. Num a => a -> a -> a

"Give me two values of some numeric type a and I'll give you back another value of that same numeric type"

In [79]:
:t (+)

(+) :: forall a. Num a => a -> a -> a

In [80]:
:t even

even :: forall a. Integral a => a -> Bool

In [81]:
:t (||)

(||) :: Bool -> Bool -> Bool

In [82]:
:t not

not :: Bool -> Bool

In [83]:
:t (==)

(==) :: forall a. Eq a => a -> a -> Bool

Hopefully this makes sense based on our other examples, but to be clear, it basically says "give me two values of some type `a` that is comparable for equality (ie. is an instance of the `Eq` class) and I will give you back a `Bool` (indicating whether they are equal)".

In [84]:
:t map

map :: forall a b. (a -> b) -> [a] -> [b]

Map is our first higher order function -- it says "give me a function from `a` to `b` and a list of `a`s and I'll give you back a list of `b`s".

Logically this makes sense -- if you have a function that can turn any `a` into a `b`, and you have a list of `a`s then you can turn it into a list of `b`'s by applying the function to each element of the list.

We've seen type variables in type signatures before but this is the first type we've seen them without any accompanying class constraints.  This just means that there are no limits on what the types `a` and `b` can be whatsoever.  

Signatures that contain type variables with class constraints are examples of "ad-hoc polymorphism" whereas signatures with no constraints are examples of "parameteric polymorphism".

We'll come back to why this is import in a moment.

In [85]:
:t filter

filter :: forall a. (a -> Bool) -> [a] -> [a]

Similarly the signature for `filter` says "give me predicate on `a`'s (a function from `a` to `Bool`) and a list of `a`'s and I'll give you back a (filtered) list of `a`'s".

In [86]:
:t foldl'

foldl' :: forall (t :: * -> *) b a. Foldable t => (b -> a -> b) -> b -> t a -> b

The signature for `foldl'` is a little scarier but basically says "given a function of type `b -> a -> b` and a single (seed value) of type `b` and a `Foldable` of `a`'s I'll give you back a single `b`.

`Foldable` is essentially a generalization of lists that allows the `foldl'` function to operate on more than just lists.  The `(t :: * -> *)` bit is redundant in this case because all `Foldables` have this shape, but basically says "where the type `t` is really a type constructor that needs to be supplied with some other concrete type in order to produce a concrete type of it's own.

For example the list type `[a]` is requires a concrete type for `a` (say `Int`) to become the concrete type `[Int]`.

In [87]:
:t Just 5

Just 5 :: forall a. Num a => Maybe a

In [88]:
:t Nothing

Nothing :: forall a. Maybe a

The type signature for both of these should mostly make sense by now but the important thing to note is that the optionality of the type is reflected in the type signature.  If there's no `Maybe` then you know the value will always be present and you never need to do any checks for a value being missing.  Likewise if there is a `Maybe` then the compiler will force you to handle both cases.

## User Defined Data Types (Round 2)

So far the only data type we've defined is our `User` "record" type from above.  Calling this was a record type is actually not really accurate.  It's technically an algebraic data type, specifically a sum type, which is declared using "record syntax".  The curly braces containing a list of name/type pairs comprises "record syntax".  This syntax is essentially sugar for declaring an ADT with the same set of fields and type as well as generating the accessor functions for each of the fields at the same time.

We could create a type almost the same as our `User` type from above like so:

In [89]:
data User' = User' Int String Bool
  deriving (Eq, Ord, Read, Show)
  
userId' :: User' -> Int
userId' (User' uid _ _) = uid

userName' :: User' -> String
userName' (User' _ n _) = n

isAdmin' :: User' -> Bool
isAdmin' (User' _ _ a) = a

... but obviously it's lot nicer to write:

    data User = User
      { userId   :: Int
      , username :: String
      , isAdmin  :: Bool
      } deriving (Eq, Ord, Read, Show)

We've seen how to access the values of various fields of a `User` by using the accessor functions:

In [90]:
username adminUser

"admin"

...but we might also want to update a field, right? Well... yes, we might want to, but here we run afoul of Haskell's pervasive immutability.  There *are* various mechanisms for either emulating state (eg the `State Monad`) or achieving the experience of mutability from pure code (eg. `IORef`s, `MVar`s, `TVars`) or acheiving actual mutability (eg. the `St Monad`'s state threads) but most idiomatic Haskell code will eschew these mechanisms unless they are absolutely necessary for some reason.

How can we get anything done if we can't mutate anything?  Easy... we make copies!

Record update syntax allows us to create a new copy of a record with one or more fields updated:

In [91]:
normUser' = normUser { userId = 2, username = "Bob" }

normUser'

User {userId = 2, username = "Bob", isAdmin = False}

Our original `normUser` value still exists unchanged:

In [92]:
normUser

User {userId = 1, username = "norm", isAdmin = False}

You might be thinking "if we copy everything all of the time isn't that really expensive and inefficient?"  The short answer is "No -- at least it's not nearly as bad as you might think, thanks to..."

## Pesistent Data Structures

One primary mitigating factors is that data structures in Haskell are "persistent".  In the context of data structures (as opposed to the context of databases), "persistent" means that they are implemented in such a way as to promote sharing large chunks of themselves on copy.  

Going back to our linked list example, if we go back to our list `xs`:

In [93]:
xs

[1,2,3]

We can prepand `0` to the list like so:

In [94]:
xs' = 0:xs

xs'

[0,1,2,3]

Of course our original list `xs` is unchanged:

In [95]:
xs

[1,2,3]

but in memory, the list `xs'` is basically represented as `0:xs` -- that is to say the initial list with a pointer to the "rest of the list" which in this case is exactly `xs`.  Since neither is mutable we don't have to worry about changes to one affecting the other.  

Other data structures in Haskell share these same properties -- for example replacing the root node in a binary tree might involve creating a new node with left and right pointers pointing to the left and right subtrees of the original tree.  In this case no matter how big the original tree is, only a single new node has to be allocated for the "replace the root node" operation.  

This sharing can be a source of space leaks which are one of the more annoying parts of Haskell but (at least in my experience) they don't tend to come up nearly as much in practice as people fear they might.

## Typeclasses

Typeclasses bear no relation to OO classes.  They are closer to something like an interface in Java, though even that is a gross simplification.

In any event Typeclasses provide a mechanism for abstracting over a common set of behaviors.  We can redo our speaking Animal's example from above in a much cleaner way with typeclasses:

In [96]:
class ClassyAnimal a where
  name  :: a -> String
  speak :: a -> IO ()
  
data Dog = Dog { unDogName :: String }
data Cat = Cat { unCatName :: String }

instance ClassyAnimal Dog where
  name       = unDogName
  speak self = putStrLn (name self ++ " says Woof!")

instance ClassyAnimal Cat where
  name       = unCatName
  speak self = putStrLn (name self ++ " says Meow!")
  
speak (Dog "Otis")  
speak (Cat "Milo")

Otis says Woof!

Milo says Meow!

As you can see, this is much cleaner than our original attempt to emulate OO above.

But still, the bulk of the use of typeclasses is not for emulating OO but rather for generalizing common patterns, quite often in a "lawful" way.  A very common example of this type of abstraction is the pair of classes `Semigroup` and `Monoid`.  

A semigroup is a type paired with a binary operation for combining two elements of the type to get another element of the same type.  A monoid is a type paired with binary operation for combining two elements of the type to get another element of the same type and an identity element.  In other words, `Semigroup` is logically a superclass (in the Haskell sense, not the OO sense) of `Monoid`.

Until recently this was not reflected in the Haskell types but as of sometime last year the relationship was made explicit so `Semigroup` is actually a superclass of `Monoid` in recent versions of GHC.

In [97]:
:opt no-pager

In [98]:
import Data.Semigroup

:info Semigroup

class Semigroup a where
  (<>) :: a -> a -> a
  sconcat :: GHC.Base.NonEmpty a -> a
  stimes :: Integral b => b -> a -> a
  {-# MINIMAL (<>) #-}
  	-- Defined in ‘GHC.Base’
instance Monoid m => Semigroup (WrappedMonoid m) -- Defined in ‘Data.Semigroup’
instance Semigroup a => Semigroup (Option a) -- Defined in ‘Data.Semigroup’
instance Ord a => Semigroup (Min a) -- Defined in ‘Data.Semigroup’
instance Ord a => Semigroup (Max a) -- Defined in ‘Data.Semigroup’
instance Semigroup (Last a) -- Defined in ‘Data.Semigroup’
instance Semigroup (First a) -- Defined in ‘Data.Semigroup’
instance Semigroup (Either a b) -- Defined in ‘Data.Either’
instance Num a => Semigroup (Sum a) -- Defined in ‘base-4.12.0.0:Data.Semigroup.Internal’
instance Num a => Semigroup (Product a) -- Defined in ‘base-4.12.0.0:Data.Semigroup.Internal’
instance Semigroup (Endo a) -- Defined in ‘base-4.12.0.0:Data.Semigroup.Internal’
instance Semigroup a => Semigroup (Dual a) -- Defined in ‘base-4.12.0.0:Data.Semigroup.Internal’
instance Semigroup Any -- Defined in ‘base-4.12.0.0:Data.Semigroup.Internal’
instance Semigroup All -- Defined in ‘base-4.12.0.0:Data.Semigroup.Internal’
instance Semigroup [a] -- Defined in ‘GHC.Base’
instance Semigroup Ordering -- Defined in ‘GHC.Base’
instance Semigroup a => Semigroup (Maybe a) -- Defined in ‘GHC.Base’
instance Semigroup a => Semigroup (IO a) -- Defined in ‘GHC.Base’
instance Semigroup b => Semigroup (a -> b) -- Defined in ‘GHC.Base’
instance (Semigroup a, Semigroup b, Semigroup c, Semigroup d, Semigroup e) => Semigroup (a, b, c, d, e) -- Defined in ‘GHC.Base’
instance (Semigroup a, Semigroup b, Semigroup c, Semigroup d) => Semigroup (a, b, c, d) -- Defined in ‘GHC.Base’
instance (Semigroup a, Semigroup b, Semigroup c) => Semigroup (a, b, c) -- Defined in ‘GHC.Base’
instance (Semigroup a, Semigroup b) => Semigroup (a, b) -- Defined in ‘GHC.Base’
instance Semigroup () -- Defined in ‘GHC.Base’

I'm not sure why those display so differently, but the first 4 lines of Semigroup show the structure of the class.  I believe `sconcat` and `stimes` are essentially hooks for providing optimized versions of common `Semigroup` operations (smushing together a non-empty list of `Semigroup` values or smushing together N copies of the same `Semigroup` value, respectively) and either way we'll ignore them for now.

The key operation is `<>` which is for combining any two of the Semigroup elements.  Examples of Semigroups are:

- `First` which gives you back the leftmost item
- `Max` which gives back the biggest item
- `Min` which gives back the smallest item

In [99]:
print (First 'a' <> First 'b' <> First 'c')
print (Max 1 <> Max 55 <> Max 4)
print (Min 22 <> Min 18 <> Min 45)

First {getFirst = 'a'}

Max {getMax = 55}

Min {getMin = 18}

These are not particularly useful in cases like these isolated examples because there's often a specialized function for the common cases that makes more sense, eg. `head` for getting the first element of a list, `maximum` for getting the largest item from a list, `minimum` for getting the smallest item in a list, etc.

But they become more useful with you define an operation over any monoid and then choose the behavior you want by choosing the appropriate `Semigroup` instance.

Another example of a semigroup that is not a monoid is the positive integers with addition as the binary operation (lacking zero they cannot form a monoid).

As we said, monoids are semigroups with an identity element.  Two of the most common examples are:

- (ℕ,+,0) - the integers with addition as the binary operation and zero as the identity (ℕ,+,0)
- (ℕ,*,1) - the integers with multiplication as the binary operation and one as the identity element 

In Haskell we can only have one instance of a class per type, so we can't express both of these directly on integers.

The usual solution to this is to create newtype wrappers.  Newtype wrappers give us an extra constructor to wrap around a type to distinguish, in this case, which Monoid instance we want to use.  The two monoids we described above are represented by the `Sum` and `Product` monoid respectively:

In [100]:
print (Sum 1)
print (Sum 2)
print (getSum (Sum 3))
print (getSum (Sum 4))
print (Product 5)
print (Product 6)
print (getProduct (Product 5))
print (getProduct (Product 6))

Sum {getSum = 1}

Sum {getSum = 2}

3

4

Product {getProduct = 5}

Product {getProduct = 6}

5

6

Not particularly useful until we start combining them:

In [101]:
print (getSum (Sum 1 <> Sum 2))
print (getProduct (Product 5 <> Product 6))

3

30

It still seems like a lot of work to accomplish the same thing as  `print (1 + 1)` and `print (5 * 6)` but the benefits start to accrue when we are able to build functions that are generic over all monoids.  An example of such a function is `mconcat` which combines a list of monoidal values with their binary operation:

In [102]:
print (mconcat . map Sum     $ [1..10])
print (mconcat . map Product $ [1..10])

Sum {getSum = 55}

Product {getProduct = 3628800}

Strings (with the empty string as the identity element) are monoids, so we can `mconcat` them as well:

In [103]:
print (mconcat ["foo", "bar", "baz"])

"foobarbaz"

## Function Composition

One last thing that we haven't covered that will be useful for the next section is function composition.  Function composition works the same way it does in mathematics where 

$$ (f \circ g)  x = f ( g ( x ) )  $$

and $$ (f \circ g) x $$ can be thought of as "apply g to x then apply f to the result" except that in haskell we use a period as the dot operator.

This is a little backwards from the way you might be used methods chaining in OO languages where `X.foo.bar` means "apply foo to X then apply bar to the result" but you get used to it.

The types of the functions you are composing must line up.  A simple case would be composing a series of mathematical operations:

In [109]:
add5mul3add2 = (+2) . (*3) . (+5)

-- The types of these functions are all the same:
:t (+2)
:t (*3)
:t (+5)

add5mul3add2 10

(+2) :: forall a. Num a => a -> a

(*3) :: forall a. Num a => a -> a

(+5) :: forall a. Num a => a -> a

47

An example where the types change would be:

In [110]:
doubleTheLength = (*2) . length

:t (*2)
:t length
:t doubleTheLength

doubleTheLength "some string"

(*2) :: forall a. Num a => a -> a

length :: forall (t :: * -> *) a. Foldable t => t a -> Int

doubleTheLength :: forall (t :: * -> *) a. Foldable t => t a -> Int

22

Examining the type of the composition operator helps make it clear what's going on:

In [112]:
:t (.)

(.) :: forall b c a. (b -> c) -> (a -> b) -> a -> c

The (->) in Haskell type signatures is right associative, and we can drop the explicit `forall` for clarity, so the above is equivalent to:

```(.) :: (b -> c) -> (a -> b) -> (a -> c)```

which can be read as "given a function from `b` to `c`, and a function from `a` to `b` I'll give you back a function from `a` to `c` (by sticking the two functions you gave me together end to end)."

And that's going to have to do it for now.  Onward to our final section...